nixos/modules/services/databases/lldap.nix

```diff
@@ -0,0 +1,121 @@
+      type = types.nullOr types.path;
+      default = null;
+      description = lib.mdDoc ''
+        Environment file as defined in {manpage}`systemd.exec(5)` passed to the service.
+      '';
+    };
+
+    settings = mkOption {
+      description = mdDoc ''
+        Free-form settings written directly to the `lldap_config.toml` file.
+        Refer to <https://github.com/lldap/lldap/blob/main/lldap_config.docker_template.toml> for supported values.
+      '';
+
+      default = { };
+
+      type = types.submodule {
+        freeformType = format.type;
+        options = {
+          ldap_host = mkOption {
+            type = types.str;
+            description = mdDoc "The host address that the LDAP server will be bound to.";
+            default = "::";
+          };
+
+          ldap_port = mkOption {
+            type = types.port;
+            description = mdDoc "The port on which to have the LDAP server.";
+            default = 3890;
+          };
+
+          http_host = mkOption {
+            type = types.str;
+            description = mdDoc "The host address that the HTTP server will be bound to.";
+            default = "::";
+          };
+
+          http_port = mkOption {
+            type = types.port;
+            description = mdDoc "The port on which to have the HTTP server, for user login and administration.";
+            default = 17170;
+          };
+
+          http_url = mkOption {
+            type = types.str;
+            description = mdDoc "The public URL of the server, for password reset links.";
+            default = "http://localhost";
+          };
+
+          ldap_base_dn = mkOption {
+            type = types.str;
+            description = mdDoc "Base DN for LDAP";
+            default = "dc=example,dc=com";
+          };
+
+          ldap_user_dn = mkOption {
+            type = types.str;
+            description = mdDoc "Admin username";
+            default = "admin";
+          };
+
+          ldap_user_email = mkOption {
+            type = types.str;
+            description = mdDoc "Admin email";
+            default = "admin@example.com";
+          };
+
+          database_url = mkOption {
+            type = types.str;
+            description = mdDoc "Database URL";
+            default = "sqlite://./users.db?mode=rwc";
+            example = "postgres://postgres-user:password@postgres-server/my-database";
+          };
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.lldap = {
+      description = "Lightweight LDAP server (lldap)";
+      after = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = "${lib.getExe cfg.package} run --config-file ${format.generate "lldap_config.toml" cfg.settings}";
+        StateDirectory = "lldap";
+        WorkingDirectory = "%S/lldap";
+        User = "lldap";
+        Group = "lldap";
+        DynamicUser = true;
```

Comment:
DynamicUser does not work well with EnvironmentFile

Reply:
Care to elaborate?
Because I do use `DynamicUser` with `EnvironmentFile`s in quite a few places without any issues.

Reply:
I am not sure how `DynamicUser` handles a home directory. We need to make sure that `sqlite` database is persisted, since it is the default.

Reply:
The working directory is set to `/var/lib/lldap` via `WorkingDirectory = "%S/lldap";`.
The sqlite is relative to the working directory due to the relative path in `"sqlite://./users.db?mode=rwc"`.

You can validate this yourself using the following [`nixos-shell`](https://github.com/Mic92/nixos-shell)

```nix
{ ... }:

{
  services.lldap.enable = true;
  services.openssh.enable = true;

  virtualisation = {
    forwardPorts = [
      { from = "host"; host.port = 2222; guest.port = 22; }
    ];
  };
  nixos-shell.mounts = {
    mountHome = false;
    mountNixProfile = false;
  };
}
```

```
❯ nix-shell -p nixos-shell --command "nixos-shell -I nixpkgs=. vm.nix"
```
to start the `nixos-shell` and then executing within it:
```
[root@nixos:~]# ls -lash /var/lib/lldap/
total 64K
4.0K drwxr-xr-x 2 lldap lldap 4.0K Apr 25 19:45 .
4.0K drwx------ 3 root  root  4.0K Apr 25 19:45 ..
4.0K -r-------- 1 lldap lldap  128 Apr 25 19:45 server_key
 52K -rw-r--r-- 1 lldap lldap  48K Apr 25 19:45 users.db
```

Reply:
After the discussion on matrix, I am under the impression that `DynamicUser` has no security issues here, as the EnvironmentFile is read by systemd itself.

There's an open debate about the ergonomics of `DynamicUser` while debugging, but that is best done outside this PR as we are already using `DynamicUser` elsewhere in nixpkgs.

Reply:
> DynamicUser does not work well with EnvironmentFile

This has been resolved by a discussion over in matrix at `#systemd:nixos.org`[^1][^2]

---

@LongerHV 

[^1]: https://logs.nixos.dev/room/!DBFhtjpqmJNENpLDOv:nixos.org/?anchor=$fL3B7t12f6ChTkGLDQ4Hfxf19gOhunNwk0P7ieFcn5k&offset=26
[^2]: https://matrix.to/#/!DBFhtjpqmJNENpLDOv:nixos.org/$DFn18I9A_VqnVWPmoulBQ3y-uK-HxP6hbKSmlwtmPb8?via=nixos.org&via=matrix.org&via=nixos.dev


> I am not sure how `DynamicUser` handles a home directory. We need to make sure that `sqlite` database is persisted, since it is the default.

Are you okay with the [response I gave](https://github.com/NixOS/nixpkgs/pull/227916#discussion_r1176971904)?
I would like to mark this as resolved, if that's okay with you :)

Reply:
It is fine for me, thanks for a detailed explanation ;)